accounts/models.py

```diff
@@ -350,55 +381,123 @@ def num_sounds_pending_moderation(self):
+        data but keeping related user generated content, etc. This method handles all of that and should be always
+        used instead of User.delete().
+
+        When deleting a user using Profile.delete_user(), a DeletedUser object will be always created to leave a
+        trace of when was the user deleted and what was the reason. The original User object will be preserved but all
+        the personal data will be anonymized, and the user will be flagged as having been deleted.
+
+        When deleting a user, it's sounds and packs can either be deleted or preserved (see function args description
+        below). Other related content (ratings, comments, posts) will be preserved (but appear under a "deleted user"
+        account).
 
-        This method anonymise the user and flags it as deleted. If
-        remove_sounds is True then the Sound (and Pack) object is removed from
-        the database.
+        Optionally, a user can be fully deleted from DB including all of its packs, sounds and other relaed content.
+        Even in this case a DeletedUser object will be created to keep a record.
+
+        Args:
+            remove_sounds (bool): if True the sounds created by the user will be deleted as well. Otherwise the sounds
+              will still be available to other users but appear under a "deleted user" account. Defaults to False.
+            delete_user_object_from_db (bool): if True the user object will be completely removed from the DB together
+              with all related content. Defaults to False.
+            deletion_reason (str): reason for the user being deleted. Should be one of the choices defined in
+              DeletedUser.DELETION_REASON_CHOICES. Defaults to DeletedUser.DELETION_REASON_DELETED_BY_ADMIN.
         """
 
-        self.user.username = 'deleted_user_%s' % self.user.id
-        self.user.email = 'deleted_user_%s@freesound.org' % self.user.id
-        self.has_avatar = False
-        self.is_deleted_user = True
-        self.user.set_unusable_password()
-
-        self.about = ''
-        self.home_page = ''
-        self.signature = ''
-        self.geotag = None
-
-        self.save()
-        self.user.save()
-        if remove_sounds:
-            Sound.objects.filter(user=self.user).delete()
-            Pack.objects.filter(user=self.user).update(is_deleted=True)
-        else:
-            Sound.objects.filter(user=self.user).update(is_index_dirty=True)
+        # Run all deletion operations in a single transaction so if there is an error we don't create duplicate
+        # DeletedUser objects
+        with transaction.atomic():
+
+            # Create a DeletedUser object to store basic information for the record (first check if it
+            # already exists because user was deleted previously but db object preserved
+            try:
+                deleted_user_object = DeletedUser.objects.get(user=self.user)
```

Comment:
I think I did not do it because in the "create" case I have to provide more info and I don't know if `get_or_create` will use all the fields I pass in the query for deciding if an object actually exists. Also I think this way the code is quite clear in this case.

Reply:
this is in the [documentation for get_or_create](https://docs.djangoproject.com/en/3.0/ref/models/querysets/#get-or-create) - you pass in the filter objects as regular parameters, and pass in new fields to fill in as the `defaults` parameter, so it's possible to do this behaviour. No problem